### PR TITLE
perf(813): Avoid materialization of product values in subscriptions

### PR DESCRIPTION
**Pull request by Centril**
# Description of Changes

- Fixes https://github.com/clockworklabs/SpacetimeDB/issues/813.

```
Benchmarking full-scan: Collecting 100 samples in estimated 
full-scan               time:   [162.84 ms 164.37 ms 166.88 ms]
                        change: [-47.765% -47.089% -46.260%] (p = 0.00 < 0.05)
                        Performance has improved.

Benchmarking query-indexes-multi: Collecting 100 samples in 
query-indexes-multi     time:   [1.1162 µs 1.1202 µs 1.1274 µs]
                        change: [-6.4601% -5.3782% -4.0002%] (p = 0.00 < 0.05)
                        Performance has improved.
```

# API and ABI breaking changes

None

# Expected complexity level and risk

2

### CODE REVIEW

**Comment by joshua-spacetime** [comment]
Here are the numbers I get compared to latest master:
```
full-scan               time:   [82.200 ms 82.378 ms 82.581 ms]
                        change: [-16.491% -16.056% -15.598%] (p = 0.00 < 0.05)
                        Performance has improved.

full-join               time:   [236.58 µs 236.81 µs 237.07 µs]
                        change: [-8.2742% -8.0187% -7.7901%] (p = 0.00 < 0.05)
                        Performance has improved.

incr-select             time:   [666.93 ns 667.28 ns 667.64 ns]
                        change: [+53.585% +53.879% +54.170%] (p = 0.00 < 0.05)
                        Performance has regressed.

incr-join               time:   [3.6968 µs 3.7051 µs 3.7123 µs]
                        change: [-18.685% -18.431% -18.173%] (p = 0.00 < 0.05)
                        Performance has improved.
```

I haven't reviewed this patch yet, so I don't know where the regression in `incr-select` is coming from. I'll get back to you once I have more information. But pretty good numbers otherwise.

**Comment by jdetter** [comment]
Master benchmark on 14900k:
```
    Finished bench [optimized + debuginfo] target(s) in 0.15s
     Running benches/subscription.rs (/home/ubuntu/SpacetimeDB/target/release/deps/subscription-7c24f5921892496c)
Benchmarking full-scan: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 29.7s, or reduce sample count to 10.
full-scan               time:   [243.00 ms 243.98 ms 244.95 ms]

full-join               time:   [686.00 µs 686.16 µs 686.33 µs]
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) low mild
  1 (1.00%) high severe

incr-select             time:   [481.07 ns 481.26 ns 481.45 ns]
Found 6 outliers among 100 measurements (6.00%)
  3 (3.00%) high mild
  3 (3.00%) high severe

incr-join               time:   [6.4075 µs 6.4135 µs 6.4199 µs]
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high mild

query-indexes-multi     time:   [925.65 ns 925.95 ns 926.25 ns]
```

crntril/borrow-eq:
```
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 12.0s, or reduce sample count to 40.
full-scan               time:   [113.10 ms 113.30 ms 113.50 ms]
                        change: [-53.768% -53.563% -53.363%] (p = 0.00 < 0.05)
                        Performance has improved.

full-join               time:   [573.18 µs 573.49 µs 573.83 µs]
                        change: [-16.468% -16.422% -16.372%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) low mild
  1 (1.00%) high mild

incr-select             time:   [455.01 ns 455.35 ns 455.72 ns]
                        change: [-5.4337% -5.3479% -5.2623%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high mild

incr-join               time:   [5.0606 µs 5.0649 µs 5.0692 µs]
                        change: [-21.097% -21.008% -20.923%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

query-indexes-multi     time:   [843.18 ns 843.48 ns 843.79 ns]
                        change: [-8.9223% -8.8640% -8.8035%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 10 outliers among 100 measurements (10.00%)
  7 (7.00%) high mild
  3 (3.00%) high severe
```

**Review by joshua-spacetime — CHANGES_REQUESTED** [review]
The performance numbers are good, so I would like to merge this as soon as possible.

However I'm particularly concerned about the extra lifetime that was added to the datastore traits. I'm totally fine with it, if the perf justifies it. But I am not aware of it being an issue, and so I really think we should revert those changes.

**Comment by cloutiertyler** [comment]
I, too, am a bit concerned about modifying the datastore trait in this way, if only because it may have implications on MVCC which I need to be aware of. Could you please make a post here (or better yet a comment in the codebase) which explains the purpose of this lifetime and why it is necessary?

**Comment by gefjon** [comment]
I'd really like for these two tickets to get separate PRs so we can measure the performance impact of the solutions separately.

File: crates/core/src/vm.rs [hunk 1]
```diff
@@ -247,25 +249,25 @@ fn iter_by_col_range<'a>(
...
+    pub index_header: &'c Arc<Header>,
     /// An optional predicate to evaluate over the matching rows of the index.
-    pub index_select: Option<ColumnOp>,
+    pub index_select: &'c Option<ColumnOp>,
     /// The table id on which the index is defined.
     pub index_table: TableId,
-    /// The column id for which the index is defined.
-    pub index_col: ColId,
+    /// The column ids for which the index is defined.
+    pub index_cols: &'c ColList,
```
[the remark this answers is not in the document]
Comment (joshua-spacetime) [reply crates/core/src/vm.rs:264]: This is probably something I would revert. I think it's fine to keep changes related to #813 and #833 in this patch, but this change, that is, passing a borrowed ColList, if we are going to make it, probably should be in a separate patch.

File: crates/core/src/subscription/subscription.rs [hunk 2]
```diff
@@ -491,87 +498,101 @@ impl IncrementalJoin {
...
 
         join_5.retain(|row| !join_6.remove(row));
 
-        Ok(Some(
-            join_2
-                .into_iter()
-                .chain(join_4)
-                .chain(join_6)
-                .map(TableOp::delete)
-                .chain(join_1.into_iter().chain(join_5).map(TableOp::insert)),
```
[the remark this answers is not in the document]
Comment (joshua-spacetime) [reply crates/core/src/subscription/subscription.rs:567]: What temporary allocations exactly? These are iterators so we should be able to chain them still right?

**Comment by joshua-spacetime** [comment]
```
full-scan               time:   [84.373 ms 84.635 ms 84.909 ms]
                        change: [-16.825% -16.324% -15.837%] (p = 0.00 < 0.05)
                        Performance has improved.

full-join               time:   [241.59 µs 242.01 µs 242.56 µs]
                        change: [-13.306% -12.923% -12.537%] (p = 0.00 < 0.05)
                        Performance has improved.
```

**Review by joshua-spacetime — APPROVED** [review]

**Pull request merged**